File: Es_Python/Prove/ProxySkeleton/skeleton.py
from serverInterface import ServerInterface
import socket

import threading

class Skeleton(ServerInterface):
    """
    parametri: port, server_impl
    """
    
    def __init__(self, port, server_impl):
        
        self.port = port
        self.server_impl = server_impl
        
    def inverti_stringa(self, data):
        print("<SKELETON> inoltro a inverti_stringa del delegato...")
        r = self.server_impl.inverti_stringa(data)
        
        return r
    
    def run_skeleton(self):
        
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s.bind(('localhost', self.port))
        
        s.listen()
        
        print("<SERVER SKELETON> in ascolto su porto: ", self.port)
        
        while True:
            conn, addr = s.accept()
            
            #affido tutto al thread
            t = threading.Thread(target = run_function, args=(conn, self))
            
            t.start()
        
        s.close()
        
def run_function(conn:socket.socket, skeleton:Skeleton):
    
    richiesta = conn.recv(1024).decode("utf-8")
    
    risposta = skeleton.inverti_stringa(richiesta)
    
    #invia messaggio di risposta al client
    conn.send(risposta.encode("utf-8"))
    
    conn.close()
    

File: Es_Java/Esercitazioni/ProxySkeleton/Prova1_TCP_Delega/src/Skeleton.java
import java.io.IOException;
import java.net.*;

public class Skeleton implements IDispatcher{
    IDispatcher delegato;
    int port;

    public Skeleton(IDispatcher del, int port){
        this.delegato = del;
        this.port=port;
    }

    public void sendCmd(int command){
        delegato.sendCmd(command);
    }

    public int getCmd(){
        return delegato.getCmd();
    }

    public void run_skeleton(){
        //Avvia il server
        try(ServerSocket s = new ServerSocket(port)){
            System.out.print("Server listening on "+s.getLocalSocketAddress()+" and port: "+s.getLocalPort());
            
            while (true) {
                Socket conn = s.accept();
                System.out.println("Ricevuta richiesta di connessione, passo a thread...");

                ServerThread td = new ServerThread(conn, this);
                td.start();

            }
        }
        catch(IOException e){
            e.printStackTrace();
        }

    }
}
